src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs

```diff
@@ -278,13 +278,57 @@ protected Task ResolveAuthorityAsync()
             string scopes = GetOverridenScopes(AuthenticationRequestParameters.Scope).AsSingleString();
             var tokenClient = new TokenClient(AuthenticationRequestParameters);
 
+            var CCSHeader = GetCCSHeader(additionalBodyParameters);
+            if (CCSHeader != null)
+            {
+                tokenClient.AddHeaderToClient(CCSHeader.Item1, CCSHeader.Item2);
+            }
+
             return tokenClient.SendTokenRequestAsync(
                 additionalBodyParameters,
                 scopes,
                 tokenEndpoint,
                 cancellationToken);
         }
 
+        //The CCS header is used by the CCS service to help route requests to resources in Azure during requests to speed up authentication.
+        //It consists of either the ObjectId.TenantId or the upn of the account signign in.
+        //See https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2525
+        protected virtual Tuple<string, string> GetCCSHeader(IDictionary<string, string> additionalBodyParameters)
+        {
+            if (AuthenticationRequestParameters?.Account?.HomeAccountId != null)
+            {
+                if (!String.IsNullOrEmpty(AuthenticationRequestParameters.Account.HomeAccountId.Identifier))
+                {
+                    var userObjectId = AuthenticationRequestParameters.Account.HomeAccountId.ObjectId;
+                    var userTenantID = AuthenticationRequestParameters.Account.HomeAccountId.TenantId;
+                    string OidCCSHeader = CoreHelpers.GetCCSClientInfoheader(userObjectId, userTenantID);
+
+                    return new Tuple<string, string>(Constants.OidCCSHeader, OidCCSHeader);
+                }
+                else if (!String.IsNullOrEmpty(AuthenticationRequestParameters.Account.Username))
```

Comment:
Please simplify these if / else clauses ... for example since you return at line 307, you don't need the `else`.